rwengine/src/engine/GarageController.cpp

```diff
@@ -0,0 +1,702 @@
+#include "GarageController.hpp"
+
+#include <btBulletDynamicsCommon.h>
+#include <glm/gtx/quaternion.hpp>
+
+#include "dynamics/CollisionInstance.hpp"
+
+#include "ai/PlayerController.hpp"
+
+#include "engine/GameState.hpp"
+
+#include "objects/CharacterObject.hpp"
+#include "objects/GameObject.hpp"
+#include "objects/InstanceObject.hpp"
+#include "objects/VehicleObject.hpp"
+
+GarageController::GarageController(GameWorld* engine, GarageInfo* info,
+                                   InstanceObject* door)
+
+    : swingType(false)
+    , needsToUpdate(false)
+    , fraction(0.f)
+    , step(0.5f)
+    , doorHeight(4.0f)
+    , engine(engine)
+    , garageInfo(info)
+    , doorObject(door) {
+    if (doorObject) {
+        startPosition = doorObject->getPosition();
+        startRotation = doorObject->getRotation();
+    }
+    if (garageInfo) {
+        if (garageInfo->state == GarageState::Closed) {
+            fraction = 0.f;
+        } else {
+            fraction = 1.f;
+        }
+    }
+
+    // @todo set door height according to model size
+
+    updateDoor();
+}
+
+GarageController::~GarageController() {
+}
+
+float GarageController::getDistanceToGarage(glm::vec3 point) {
+    float dx = std::max(
+        {garageInfo->min.x - point.x, 0.f, point.x - garageInfo->max.x});
+    float dy = std::max(
+        {garageInfo->min.y - point.y, 0.f, point.y - garageInfo->max.y});
+
+    // Seems like original game ignores z axis, bug or feature?
+    // float dz = std::max(
+    //    {garageInfo->min.z - point.z, 0.f, point.z - garageInfo->max.z});
+
+    // return std::sqrt(dx * dx + dy * dy + dz * dz);
+
+    return std::sqrt(dx * dx + dy * dy);
+}
+
+bool GarageController::isObjectInsideGarage(GameObject* object) {
+    // This is not that trivial, we need to check if full vehicle body is inside
+    auto p = object->getPosition();
+
+    if (p.x >= (garageInfo->min.x) && p.y >= (garageInfo->min.y) &&
+        p.z >= (garageInfo->min.z) && p.x <= (garageInfo->max.x) &&
+        p.y <= (garageInfo->max.y) && p.z <= (garageInfo->max.z)) {
+        return true;
+    }
+
+    return false;
+}
+
+bool GarageController::shouldClose() {
+    auto controller = engine->players.at(0);
+    auto plyChar = controller->getCharacter();
+    auto playerPosition = plyChar->getPosition();
+    auto playerVehicle = plyChar->getCurrentVehicle();
+    bool playerIsInVehicle = playerVehicle != nullptr;
+
+    switch (garageInfo->type) {
+        case GarageType::Mission: {
+            if (!isObjectInsideGarage(static_cast<GameObject*>(plyChar)) &&
+                isObjectInsideGarage(garageInfo->target)) {
+                return true;
+            }
+
+            return false;
+        }
+
+        case GarageType::BombShop1:
```

Comment:
I think you should add in "empty cases" `break`.

Reply:
It's not empty, it falls through

Reply:
So it is intentional?

Reply:
Yes

Reply:
Bombshops and resprays logic is different though, so this is incorrect anyway. But I was advised not to add more logic as there are a lot of code already.